File: JUNHEE/프로그래머스_week5/괄호 회전하기.cpp
#include <string>
#include <vector>
#include <stack>
#include <iostream>
using namespace std;


bool is_right(string s){
    stack<char> str;
    char temp;
    for(int i=0;i<s.size();i++){
        temp = s[i];
        if(temp == ')'){
            if(str.top() == '('){
                str.pop();
            }
            else{
                str.push(temp);
            }
        }
        else if(temp == ']'){
            if(str.top() == '['){
                str.pop();
            }
            else{
                str.push(temp);
            }
        }
        else if(temp == '}'){
            if(str.top() == '{'){
                str.pop();
            }
            else{
                str.push(temp);
            }
        }
        else{
            str.push(temp);
        }
    }
    if(str.empty()) return true;
    else return false;
}
string rotation(string s){
    char start = s[0];
    for(int i=0;i<s.size()-1;i++){
        s[i] = s[i+1];
    }
    s[s.size()-1] = start;
    return s;
}
int solution(string s) {
    int answer = 0;
    is_right(s);
    for(int i = 0; i < s.size();i++){
        s = rotation(s);
        if(is_right(s))
            answer++;
    }
    return answer;
}

File: DONGUK/프로그래머스_WEEK 5/괄호 회전하기.py
from collections import deque

def check(li):
    
    stack = []
    
    for i in range(len(li)):
        top = li[i]
        
        if top == '(' or top == '{' or top == '[':
            stack.append(top)
            
        else:
            if len(stack) == 0:
                return False
            
            stack_top = stack.pop()
            if top == ')':
                if stack_top != '(':
                    return False
            elif top == '}':
                if stack_top != '{':
                    return False
            
            else:
                if stack_top != '[':
                    return False
        
    return True
    

def solution(s):
    answer = 0
    
    if len(s) % 2 == 1:
        return answer
    
    if check(s):
        answer += 1
    
    for _ in range(len(s) - 1):
        top = s[0]
        s = s[1::]
        s += top

        if check(s):
            answer += 1
        
    return answer
